lib/galaxy/jobs/runners/__init__.py

```diff
@@ -678,14 +678,19 @@ def __init__(self, job_wrapper: "JobWrapper", job_destination: "JobDestination")
         self.job_wrapper = job_wrapper
         self.job_destination = job_destination
         self.runner_state = None
-        self.exit_code_file = default_exit_code_file(job_wrapper.working_directory, job_wrapper.get_id_tag())
-
         self.redact_email_in_job_name = True
+        self._exit_code_file = None
         if self.job_wrapper:
             self.redact_email_in_job_name = self.job_wrapper.app.config.redact_email_in_job_name
 
         self.cleanup_file_attributes = ["job_file", "output_file", "error_file", "exit_code_file"]
 
+    @property
+    def exit_code_file(self) -> str:
+        return self._exit_code_file or default_exit_code_file(
+            self.job_wrapper.working_directory, self.job_wrapper.get_id_tag()
+        )
```

Comment:
Would it make sense to cache it?

```suggestion
        if not self._exit_code_file:
            self._exit_code_file = default_exit_code_file(
                self.job_wrapper.working_directory, self.job_wrapper.get_id_tag()
            )
        return self._exit_code_file
```

Reply:
`job_wrapper.working_directory` is already cached, which is the "expensive" part here, the rest is just string concatenation of attributes we've already loaded from the database.